benchmark/checkout.c

```diff
@@ -0,0 +1,193 @@
+	{ GITBENCH_OPT_SWITCH, "autocrlf",   0,   NULL,      "turn on core.autocrlf=true" },
+	{ 0 }
+};
+
+int checkout_run(gitbench_benchmark *b, gitbench_run *run)
+{
+	git_repository *repo = NULL;
+	git_oid id;
+	git_object *obj = NULL;
+	git_config *config = NULL;
+	git_checkout_options opts = GIT_CHECKOUT_OPTIONS_INIT;
+	gitbench_benchmark_checkout *benchmark = (gitbench_benchmark_checkout *)b;
+	git_buf wd_path = GIT_BUF_INIT, config_path = GIT_BUF_INIT;
+	const char *ref_name = "HEAD";
+	int error;
+
+	/* setup repository */
+	gitbench_run_start_operation(run, CHECKOUT_OPERATION_SETUP);
+
+	if (benchmark->ref_name)
+		ref_name = benchmark->ref_name;
+
+	if ((error = git_repository_open(&repo, benchmark->repo_path)) < 0 ||
+		(error = git_reference_name_to_id(&id, repo, ref_name)) < 0 ||
+		(error = git_object_lookup(&obj, repo, &id, GIT_OBJ_ANY)) < 0 ||
+		(error = git_buf_joinpath(&wd_path, run->tempdir, "workdir")) < 0 ||
+		(error = git_futils_mkdir(wd_path.ptr, NULL, 0700, GIT_MKDIR_VERIFY_DIR)) < 0 ||
+		(error = git_repository_set_workdir(repo, wd_path.ptr, 0)) < 0 ||
+		(error = git_config_new(&config)) < 0) {
+		gitbench_run_finish_operation(run);
+		goto done;
+	}
+
+	git_repository_set_index(repo, NULL);
+	git_repository_set_config(repo, config);
+
+	opts.checkout_strategy = GIT_CHECKOUT_FORCE;
+
+	gitbench_run_finish_operation(run);
+
+	if (benchmark->autocrlf) {
+		gitbench_run_start_operation(run, CHECKOUT_OPERATION_SETUP_FILTERS);
+
+		if ((error = git_buf_joinpath(&config_path, run->tempdir, ".config")) == 0 &&
+			(error = git_config_add_file_ondisk(config, config_path.ptr, GIT_CONFIG_LEVEL_LOCAL, 0)) == 0)
+			error = git_config_set_bool(config, "core.autocrlf", true);
+
+		gitbench_run_finish_operation(run);
+
+		if (error < 0)
+			goto done;
+	}
+
+	/* do the checkout */
+	gitbench_run_start_operation(run, CHECKOUT_OPERATION_CHECKOUT);
+	error = git_checkout_tree(repo, obj, &opts);
+	gitbench_run_finish_operation(run);
+
+done:
+	/* cleanup */
+	gitbench_run_start_operation(run, CHECKOUT_OPERATION_CLEANUP);
+	git_object_free(obj);
+	git_config_free(config);
+	git_repository_free(repo);
+	git_buf_free(&wd_path);
+	git_buf_free(&config_path);
+	gitbench_run_finish_operation(run);
+
+	return error;
+}
+
+void checkout_free(gitbench_benchmark *b)
+{
+	gitbench_benchmark_checkout *benchmark = (gitbench_benchmark_checkout *)b;
+
+	if (!b)
+		return;
+
+	git__free(benchmark->repo_path);
+	git__free(benchmark->ref_name);
+	git__free(benchmark);
+}
+
+int checkout_configure(
+	gitbench_benchmark_checkout *benchmark,
+	int argc,
+	const char **argv)
+{
+	gitbench_opt_parser parser;
+	gitbench_opt opt;
+
+	gitbench_opt_parser_init(&parser, checkout_cmdline_opts, argv + 1, argc - 1);
+
+	while (gitbench_opt_parser_next(&opt, &parser)) {
+		if (!opt.spec) {
+			fprintf(stderr, "%s: unknown argument: '%s'\n", progname, argv[parser.idx]);
+			gitbench_opt_usage_fprint(stderr, progname, checkout_cmdline_opts);
```

Comment:
Should we have a second version of help/usage here that effectively does ("progname cmdname") in the output?